### PR TITLE
Move files to subfolders in file directory

### PR DESCRIPTION
**Pull request by Siedlerchr**
Implementation for #1092 
ATM hardcoded to entrytype
ATM only Drag and Drop respects fileDirPattern

Already implemented Copy and Move methods from #1861 

- [x] Change in CHANGELOG.md described
- [x] Tests created for changes
- [X] Screenshots added (for bigger UI changes)
- [x] Manually tested changed features in running JabRef
- [x] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

jabreffiledirpattern

### CODE REVIEW

**Comment by Siedlerchr** [comment]
TODO: Pass global as param
TODO: Check relative path

**Comment by Siedlerchr** [comment]
@JabRef/developers  I need some help with the GUI: I desperately tried to add the new field below the other, but I can't get it work, no matter what I tried :rage: 

The original layout code is this: 
`FormLayout layout = new FormLayout("1dlu, 8dlu, left:pref, 4dlu, fill:3dlu")` in:
https://github.com/JabRef/jabref/blob/fileDirPattern/src/main/java/net/sf/jabref/gui/preftabs/ImportSettingsTab.java

The new field is there, but kinda hidden. No idea how to change it back:

settingsimportpanel

**Comment by Siedlerchr** [comment]
Yeah, finally the layout works :sunglasses:

**Comment by dierkes** [comment]
Thanks for the work! I'm really missing this feature in JabRef 3.

However, the current behavior is a little unexpected. First, the file seems to be moved relative to its current directory. For example directory pattern '\EntryType' moves the file with every execution of cleanup to a new subdirectory. For me, expected behavior would be, that the new path is relative to the directory of the bib-file or some otherwise specified root directory. 

Second, the file field in the bib entry seems to be adjusted only, if the option 'Use the BIB file location as primary file directory' is set. Otherwise, the file is moved, but the file path is not adjusted.

**Comment by Siedlerchr** [comment]
@dierkes  Thanks for your feedback. I would have asked you to test it otherwise, too ;)
The moving files is a bit complicated as it has to be kept backwards compatible.

The file directory is the first found directory: a) global file dire from setting b) bib file relative c) bib specific setting (atm not sure about the order b and c)

Regarding the file field, I will check it

**Comment by dierkes** [comment]
Regarding the new file directory: That's not the behavior I observed.

My test case was:
original file: phdthesis/xyz_2015.pdf (all paths relative to the bib-file)
file name pattern: '\bibtexkey'
file dir pattern: '\EntryType'

cleanup -> phdthesis/PhdThesis/xyz_215.pdf
cleanup -> phdthesis/PhdThesis/PhdThesis/xyz_215.pdf
cleanup -> phdthesis/PhdThesis/PhdThesis/PhdThesis/xyz_215.pdf

It didn't matter if 'Use the BIB file location as primary file directory' was set or if the default directory for files was set. (I tried with setting it to the directory of the bib-file.) I didn't try to set the default file directory for the specific bib-file, however. (Can do that later.)

From my perspective, a backward compatible solution could be: If the file dir pattern is not set, rename the file in its current location. If it is set, move the file to the directory specified by file dir pattern (relative to bib-file directory, if it's a relative path) and rename it there. However, an empty file name pattern could mean a) rename the file in place or b) move file to bib-file directory, so an additional option is probably necessary, that specifies whether moving files is active or not.

Interpreting the file dir pattern as a path relative to the current directory of the file doesn't make sense to me, since executing cleanup would always change the file path. Which is probably the opposite effect one expects from a cleanup ;)

**Comment by Siedlerchr** [comment]
Ah I see you are right.

Will look around.

2016-10-14 12:48 GMT+02:00 dierkes <notifications@github.com>:

**Comment by Siedlerchr** [comment]
@dierkes  I fixed it, when all tests are green you could try it out.

**Comment by dierkes** [comment]
Now it works mostly as expected for me.

Of course, as described above, with the current state it's not possible to rename a file only (if it's current location doesn't match the file dir pattern). So we might need the option to activate/deactivate moving the files.

And a very strange thing I noticed: If 'FileDir pattern' is empty and 'Main file directory' is empty and 'Use the BIB file location as primary file directory' is not set, then the file is not moved and a directory relative to the JabRef directory (or maybe to the current working directory)  is created with the name of the desired file path. Staying in the example from above, instead of renaming the file to
(literature/) PhdThesis/xyz_2015.pdf
a directory named
(jabref/) PhdThesis/xyz_2015.pdf/
is created.

**Comment by Siedlerchr** [comment]
Thanks for the feedback, I will check it
Normally renaming file only should work if fileDirPattern is empty in the
settings


Am 15.10.2016 11:52 nachm. schrieb "dierkes" <notifications@github.com>:

**Comment by Siedlerchr** [comment]
@dierkes  I fixed the strange behaviour. There was some directory path which was just "" and not null, so it was resolved to the current dir of the jar.  Now it only renames the file.
And the old behaviour now works, too, when fileDirPattern is Empty, the file is moved to the file directory, 
The file directories are checked in this order:
 The settings are prioritized in the following order and the first defined setting is used

1. metadata user-specific directory
2. metadata general directory
3. preferences directory
4. BIB file directory

**Comment by dierkes** [comment]
This works, almost. ;-)

I encountered a similar problem when dragging a file and dropping it to an entry. When I select to move the file, again instead of a file a directory is created (same path as the file should have).

**Comment by dierkes** [comment]
And another issue: If multiple files are linked to an entry, the first file is renamed and the other file links get lost. In JabRef 3.6 the first file is renamed and the other files are moved to the same directory.

**Comment by Siedlerchr** [comment]
Thanks for your intenstive testing.
The issue with the drag and drop is now fixed.  There was a problem with the layout which would have been default to bibtexkey if the pattern param is empty.

I am aware of the thing with the multi files. As far as I discovered the first successful rename/move exits the function. Will try to find out why. Not sure if I find time for it next week.

Addtionally I will create some automatic tests, too if it works finally..

**Comment by dierkes** [comment]
I think, I fixed the issue with the multiple files, see #2248.

The behavior now, as far as I understand it, is: The function will try to move/rename all attached files. If a file with the new file path already exist (that happens if two attached files have the same file extension), it doesn't move/rename the file and keeps the old file path in the bib entry.

For example, let's consider we have three attached files:
OldDir1/OldName1.pdf
OldDir2/OldName2.pdf
OldDir3/OldName3.txt

The new file paths will be:
NewDir/NewName.pdf
OldDir2/OldName2.pdf
NewDir/NewName.txt

I think, there is some room for improvement. For example, to keep OldName2 but move the file to NewDir, if possible. However, before doing this, I think it's necessary to define the desired behavior in detail.

**Comment by Siedlerchr** [comment]
Thanks for your work! Just one thing, it would be nice if you could check that locally, too.
There is a junit test I set to ignore (due to the bug) which should now work, too. Just remove the ignore-Tag
and  execute it (either directly  or run gradlew check)

` cleanupRenamePdfRenamesWithMultipleFiles()` in  `src/test/java/net/sf/jabref/logic/cleanup/CleanupWorkerTest.java `

**Comment by dierkes** [comment]
I removed the tag.

When I run `gradlew check`, the test `cleanupRenamePdfRenamesRelativeFile()` in `CleanupWorkerTest.java` fails. However, I think the problem lies within the design of the test. The expected path of the file is `Toot.tmp`, but the actual path is `Misc/Toot.tmp`. I guess, there is some default value set for fileDirPattern.

**Comment by Siedlerchr** [comment]
Hm I think the value has to be set to an empty string in the test because
otherwise it will take the value of the settings.

Am 10.11.2016 12:57 vorm. schrieb "dierkes" <notifications@github.com>:
>
> I removed the tag.
>
> When I run gradlew check, the test cleanupRenamePdfRenamesRelativeFile()
in CleanupWorkerTest.java fails. However, I think the problem lies within
the design of the test. The expected path of the file is Toot.tmp, but the
actual path is Misc/Toot.tmp. I guess, there is some default value set for
fileDirPattern.

**Comment by Siedlerchr** [comment]
@dierkes  I just fixed the failing test. As I expected it, it was due to the setting which got read

**Comment by dierkes** [comment]
Great. Is there a chance to integrate this branch into the next release?

**Comment by Siedlerchr** [comment]
Technically, this branch is based on the current master, but not included
with the 3.7 release which will be probably released today

So if you stick for the moment to the dev version of this branch, you have
all stuff from 3.7 release plus this feature.

Of course it will be in the next release then. There are only some tests
which now have to be added to make it ready for a merge

Am 13.11.2016 10:16 nachm. schrieb "dierkes" <notifications@github.com>:

File: src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java [hunk 1]
```diff
@@ -358,13 +351,33 @@ private boolean showLinkMoveCopyRenameDialog(String linkFileName, ExternalFileTy
...
 
-        renameToTextBox.setText(targetName.concat(".").concat(fileType.getExtension()));
+        String fileDirPattern = Globals.prefs.get(JabRefPreferences.IMPORT_FILEDIRPATTERN);
 
+        String targetDirName = "";
+        if (!fileDirPattern.isEmpty()) {
+            targetDirName = FileUtil.createFileNameFromPattern(database, entry, fileDirPattern, layoutPrefs);
+        }
+
+        System.out.println("TARGET Filename " + targetName); //fileNam pattern, e.g bibtexkey
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java]: Remove Sysouts, please.

**Comment by koppor** [comment]
Documentation: Could you please add an issue at help.jabref.org or even create an initial page with `skeleton: true` in the header?

You can link https://help.jabref.org/en/CustomExports#built-in-export-formatters for a description of the formatters which can be used.

File: src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java [hunk 2]
```diff
@@ -471,45 +483,42 @@ private void doLink(BibEntry entry, ExternalFileType fileType, String filename,
...
-            if (answer == JOptionPane.NO_OPTION) {
+
+            Path fromFile = Paths.get(fileName);
+            try {
+                if (!Files.exists(destFile)) {
+                    Files.createDirectories(destFile);
+                }
+            } catch (IOException e) {
+                // TODO Auto-generated catch block
+                LOGGER.error("Problem creating target direcotires", e);
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java]: directories

File: src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java [hunk 3]
```diff
@@ -471,45 +483,42 @@ private void doLink(BibEntry entry, ExternalFileType fileType, String filename,
...
+            }
+            if (FileUtil.renameFile(fromFile, destFile, true)) {
+                return true;
+            } else {
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java]: else can be omitted?!

File: src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java [hunk 4]
```diff
@@ -538,23 +548,32 @@ private boolean doCopy(String fileName, String toFile, NamedCompound edits) {
             LOGGER.warn("Cannot determine destination directory or destination directory does not exist");
             return false;
         }
-        String destinationFileName = new File(toFile).getName();
 
-        File destFile = new File(dirs.get(found) + OS.FILE_SEPARATOR + destinationFileName);
-        if (destFile.equals(new File(fileName))) {
+        Path destFile = Paths.get(dirs.get(found)).resolve(toFile);
+        System.out.println("COpy Handler dest " + destFile);
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java]: Remove Sysout

File: src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java [hunk 5]
```diff
@@ -358,13 +351,33 @@ private boolean showLinkMoveCopyRenameDialog(String linkFileName, ExternalFileTy
         moveRadioButton.setText(Localization.lang("Move file to file directory"));
         renameCheckBox.setText(Localization.lang("Rename file to").concat(": "));
 
+        //TODO: here to check for
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java]: What is the TODO here? Should it be removed?

File: src/main/java/net/sf/jabref/logic/cleanup/RenamePdfCleanup.java [hunk 6]
```diff
@@ -54,59 +64,88 @@ public RenamePdfCleanup(boolean onlyRelativePaths, BibDatabaseContext databaseCo
...
-                    newFileEntryFileName = parent.toString().concat(OS.FILE_SEPARATOR).concat(newFilename.toString());
+                    unsuccessfulRenames++;
                 }
-                newFileList.add(new ParsedFileField(description, newFileEntryFileName, type));
-            } else {
-                unsuccessfulRenames++;
             }
         }
 
+        //TODO: Checken: Der macht das return wenn change = true, aber wenn noch ein nicht change dahinter kommt?`
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/cleanup/RenamePdfCleanup.java]: There are some TODOs left?!

File: src/main/java/net/sf/jabref/logic/cleanup/RenamePdfCleanup.java [hunk 7]
```diff
@@ -120,6 +159,7 @@ public RenamePdfCleanup(boolean onlyRelativePaths, BibDatabaseContext databaseCo
         }
 
         return Collections.emptyList();
+
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/cleanup/RenamePdfCleanup.java]: Some empty lines could be removed in this PR.

File: src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java (hunk 5 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java]: Oops. Seems like I forgot that to delete. I will remove it

File: src/main/java/net/sf/jabref/logic/util/io/FileUtil.java [hunk 8]
```diff
@@ -171,11 +170,29 @@ public static boolean copyFile(Path pathToSourceFile, Path pathToDestinationFile
      * @param toFile   The target fileName
      * @return True if the rename was successful, false if an exception occurred
      */
-    public static boolean renameFile(String fromFile, String toFile) {
+    public static boolean renameFile(Path fromFile, Path toFile) {
+
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/util/io/FileUtil.java]: Empty line.

File: src/main/java/net/sf/jabref/logic/layout/format/RTFChars.java [hunk 9]
```diff
@@ -206,46 +206,46 @@ private StringInt getPart(String text, int i, boolean commandNestedInBraces) {
      * @return returns the basic character of the given unicode
      */
     private String transformSpecialCharacter(long c) {
-        if((192 <= c && c <= 197) || c == 256 || c == 258 || c == 260) {
+        if(((192 <= c) && (c <= 197)) || (c == 256) || (c == 258) || (c == 260)) {
```
Comment (koppor) [inline src/main/java/net/sf/jabref/logic/layout/format/RTFChars.java]: Could you please run the formatter on the whole method (after using `gradlew eclipse` from master branch). Then, the necessary space after `if` should be added, too.

File: src/main/java/net/sf/jabref/model/database/BibDatabaseContext.java [hunk 10]
```diff
@@ -28,6 +31,7 @@
     private DBMSSynchronizer dbmsSynchronizer;
     private DatabaseLocation location;
 
+
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/model/database/BibDatabaseContext.java]: Empty line.

File: src/main/java/net/sf/jabref/model/database/BibDatabaseContext.java [hunk 11]
```diff
@@ -154,18 +170,18 @@ public boolean isBiblatexMode() {
     * @param fieldName The field type
     * @return The default directory for this field type.
     */
-    public List<String> getFileDirectory(String fieldName, FileDirectoryPreferences preferences) {
-        List<String> fileDirs = new ArrayList<>();
+    public List<String> getFileDirectories(String fieldName, FileDirectoryPreferences preferences) {
+        List<String> fileDirs = new ArrayList<>(4);
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/model/database/BibDatabaseContext.java]: Why 4?

File: src/main/java/net/sf/jabref/model/database/BibDatabaseContext.java [hunk 12]
```diff
@@ -182,8 +198,7 @@ public boolean isBiblatexMode() {
                 fileDirs.add(parentDir);
             }
         });
-
-        return fileDirs;
+        return  fileDirs;
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/model/database/BibDatabaseContext.java]: Obsolete 2nd space?

File: src/test/java/net/sf/jabref/logic/cleanup/CleanupWorkerTest.java [hunk 13]
```diff
@@ -323,4 +327,10 @@ public void cleanupWithDisabledFieldFormatterChangesNothing() {
         worker.cleanup(preset, entry);
         Assert.assertEquals(Optional.of("01"), entry.getField("month"));
     }
+
+    @After
+    public void tearDown() {
+        JabRefPreferences.getInstance().putBoolean(JabRefPreferences.BIB_LOC_AS_PRIMARY_DIR, false);
+
```
Comment (stefan-kolb) [inline src/test/java/net/sf/jabref/logic/cleanup/CleanupWorkerTest.java]: Would be cool if you can remove all thos obsolete empty lines 😄 I know its not important but 😸

File: src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java (hunk 4 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/gui/externalfiles/DroppedFileHandler.java]: Hm, that is weird. I already deleted all that stuff I thougt.

**Comment by Siedlerchr** [comment]
I fixed the things from the review

File: src/test/java/net/sf/jabref/logic/cleanup/CleanupWorkerTest.java [hunk 14]
```diff
@@ -54,10 +55,17 @@ public void setUp() throws IOException {
         MetaData metaData = new MetaData();
         metaData.setDefaultFileDirectory(pdfFolder.getAbsolutePath());
         BibDatabaseContext context = new BibDatabaseContext(new BibDatabase(), metaData, bibFolder.newFile("test.bib"));
+
+        JabRefPreferences prefs = JabRefPreferences.getInstance();
+        prefs.putBoolean(JabRefPreferences.BIB_LOC_AS_PRIMARY_DIR, true);
```
Comment (tobiasdiez) [inline src/test/java/net/sf/jabref/logic/cleanup/CleanupWorkerTest.java]: This really changes the preferences of the developer. Can you try to mock/init by hand the layout formatter and file directory preferences.

File: src/test/java/net/sf/jabref/logic/cleanup/RenamePdfCleanupTest.java [hunk 15]
```diff
@@ -32,13 +35,16 @@
...
 
+
     @Before
     public void setUp() throws Exception {
         prefs = JabRefPreferences.getInstance();
         MetaData metaData = new MetaData();
         context = new BibDatabaseContext(new BibDatabase(), metaData, new Defaults());
         context.setDatabaseFile(testFolder.newFile("test.bib"));
 
+        prefs.putBoolean(JabRefPreferences.BIB_LOC_AS_PRIMARY_DIR, true);
```
Comment (tobiasdiez) [inline src/test/java/net/sf/jabref/logic/cleanup/RenamePdfCleanupTest.java]: as above

**Comment by Siedlerchr** [comment]
included feedbacjk

**Comment by stefan-kolb** [comment]
One or two of my comments are still open when I look at the Github UI above. Actually I'm only interested why you initialized the Array with a length of 4, I'm fine with the rest as thes are only cosmetic changes.

File: src/main/java/net/sf/jabref/model/database/BibDatabaseContext.java (hunk 11 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/model/database/BibDatabaseContext.java]: Seems like it was an artifact from my previous implementation approach.. It is no longer necessary.

**Comment by Siedlerchr** [comment]
Fixed the rest and the one test which @koppor commented out, in c9726d94686e6687af66476d26b0a1d08c61b41c works fine on travis now again.

**Comment by stefan-kolb** [comment]
@JabRef/developers we can merge right?

**Comment by lenhard** [comment]
@stefan-kolb @Siedlerchr Yes we can. That is, once the merge conflict in the italian localization gets fixed.

**Comment by Siedlerchr** [comment]
@JabRef/developers  all Conflicts fixed.

**Comment by dierkes** [comment]
Hi, on my system two tests are failing, cleanupRenamePdfRenamesWithMultipleFiles and cleanUpRenamePdfRenameFileDirectoryPatternSubDirectory. However, the problem seems to be that these and other tests do not remove the temporary files properly after execution. The behavior of the cleanup function itself is as expected, but since a file with the target path exists already (left from another test), the attached file is not renamed and the assertion fails. I'm running the tests on Fedora.

**Comment by Siedlerchr** [comment]
Thanks  for  the feedback, I  will  check  that.

Am 06.12.2016 11:45 nachm. schrieb "dierkes" <notifications@github.com>:

**Comment by Siedlerchr** [comment]
Should now be fixed. Was some Preferences problem as I already thought. Mocked it away.

**Pull request merged**